src/System.Private.ServiceModel/src/System/IdentityModel/Tokens/X509SubjectKeyIdentifierClause.cs

```diff
@@ -0,0 +1,79 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Globalization;
+using System.Security.Cryptography.X509Certificates;
+using System.ServiceModel;
+
+namespace System.IdentityModel.Tokens
+{
+    public class X509SubjectKeyIdentifierClause : BinaryKeyIdentifierClause
+    {
+        private const string SubjectKeyIdentifierOid = "2.5.29.14";
+        private const int SkiDataOffset = 2;
+
+        public X509SubjectKeyIdentifierClause(byte[] ski)
+            : this(ski, true)
+        {
+        }
+
+        internal X509SubjectKeyIdentifierClause(byte[] ski, bool cloneBuffer)
+            : base(null, ski, cloneBuffer)
+        {
+        }
+
+        private static byte[] GetSkiRawData(X509Certificate2 certificate)
+        {
+            if (certificate == null)
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull(nameof(certificate));
+
+            X509SubjectKeyIdentifierExtension skiExtension =
+                certificate.Extensions[SubjectKeyIdentifierOid] as X509SubjectKeyIdentifierExtension;
+            if (skiExtension != null)
+            {
+                return skiExtension.RawData;
+            }
+            else
+            {
+                return null;
+            }
+        }
+
+        public byte[] GetX509SubjectKeyIdentifier()
+        {
+            return GetBuffer();
+        }
+
+        public bool Matches(X509Certificate2 certificate)
+        {
+            if (certificate == null)
+                return false;
+
+            byte[] data = GetSkiRawData(certificate);
+            return data != null && Matches(data, SkiDataOffset);
```

Comment:
FWIW this would fail if anyone used a Subject Key Identifier in excess of 127 bytes (because then the prefix grows to a 3rd byte, or more).  Large values are not technically illegal, but will probably never happen.

Reply:
I'll add a comment for future generations. This issue is also there on .NET Framework so I think you are right that it will probably never happen otherwise it would have been fixed there.

---
In reply to: [274035034](https://github.com/dotnet/wcf/pull/3538#discussion_r274035034) [](ancestors = 274035034)